subxt/src/client/online_client.rs

```diff
@@ -484,3 +481,49 @@ impl Update {
         &self.metadata
     }
 }
+
+/// Helper to wait until the runtime upgrade is applied on at finalized block.
+async fn wait_for_runtime_upgrade<T: Config>(
+    client: &OnlineClient<T>,
+    runtime_version: &RuntimeVersion,
+) -> Option<Result<T::Header, Error>> {
+    use scale_value::At;
+
+    let mut block_sub = match client.backend().stream_finalized_block_headers().await {
+        Ok(s) => s,
+        Err(err) => return Some(Err(err)),
+    };
+
+    let head = loop {
+        let (block, block_ref) = match block_sub.next().await? {
+            Ok(n) => n,
+            Err(err) => return Some(Err(err)),
+        };
+
+        let key: Vec<scale_value::Value> = vec![];
+        let addr = crate::dynamic::storage("System", "LastRuntimeUpgrade", key);
+
+        let chunk = match client.storage().at(block_ref).fetch(&addr).await {
+            Ok(Some(v)) => v,
+            Ok(None) => continue,
+            Err(e) => return Some(Err(e)),
+        };
+
+        let scale_val = match chunk.to_value() {
+            Ok(v) => v,
+            Err(e) => return Some(Err(e)),
+        };
+
+        let spec_version = scale_val
+            .at("spec_version")
+            .and_then(|v| v.as_u128())
+            .expect("specVersion should exist on RuntimeVersion; qed")
+            as u32;
+
+        if spec_version >= runtime_version.spec_version {
+            break block;
+        }
+    };
+
+    Some(Ok(head))
```

Comment:
I'd return the `BlockRef` instead of the `T::Header` from this; we want to keep the `BlockRef` around for as long as we might want to fetch the block, to signal to the node to keep said block details around if it can.

`BlockRef` has ` .hash()` function that you can call as well to get the block hash :)